lib/ecto/changeset.ex

```diff
@@ -1990,9 +1990,10 @@ defmodule Ecto.Changeset do
   """
   def check_constraint(changeset, field, opts \\ []) do
     constraint = opts[:name] || raise ArgumentError, "must supply the name of the constraint"
+    constraint_string = to_string(constraint)
     message    = message(opts, "is invalid")
     match_type = Keyword.get(opts, :match, :exact)
-    add_constraint(changeset, :check, to_string(constraint), match_type, field, {message, []})
+    add_constraint(changeset, :check, constraint_string, match_type, field, {message, [constraint: {:check, constraint_string}]})
```

Comment:
implementation-wise going with our 1st proposal would mean we could replace this with:

```elixir
add_constraint(changeset, :check, constraint_string, match_type, field, message)
```

and inside`add_constraint` we could change:

```elixir
%{changeset | constraints: [%{type: type, constraint: constraint, ..., error: error} | constraints])
```

to remove `error: error` because it's redundant: it can be computed from `type` and `constraint` in the place where we convert constraint errors to changeset errors.